crypto/ube/fork_detect.c

```diff
@@ -71,29 +93,112 @@ static int init_fork_detect_madv_wipeonfork(void **addr_out) {
   // unknown |advice| values.
   if (madvise(addr, (size_t)page_size, -1) == 0 ||
       madvise(addr, (size_t)page_size, MADV_WIPEONFORK) != 0) {
-    // The mapping |addr| points to is unmapped by caller.
-    munmap(addr, (size_t)page_size);
     return 0;
   }
 
-  *addr_out = addr;
   return 1;
 }
 
+#else // defined(OPENSSL_LINUX)
+
+static int init_fork_detect_wipeonfork(void *addr, long page_size) {
+  return 0;
+}
+
+#endif // defined(OPENSSL_LINUX)
+
+
+#if defined(OPENSSL_FREEBSD) || defined(OPENSSL_OPENBSD)
+
+#include <sys/mman.h>
+#include <unistd.h>
+
+// Sometimes (for example, on FreeBSD) MAP_INHERIT_ZERO is called INHERIT_ZERO
+#if !defined(MAP_INHERIT_ZERO) && defined(INHERIT_ZERO)
+    #define MAP_INHERIT_ZERO INHERIT_ZERO
+#endif
+
+static int init_fork_detect_inheritzero(void *addr, long page_size) {
+
+  GUARD_PTR(addr);
+
+  if (minherit(addr, page_size, MAP_INHERIT_ZERO) != 0) {
+    return 0;
+  }
+
+  return 1;
+}
+
+#else // defined(OPENSSL_FREEBSD) || defined(OPENSSL_OPENBSD)
+
+static int init_fork_detect_inheritzero(void *addr, long page_size) {
+  return 0;
+}
+
+#endif // defined(OPENSSL_FREEBSD) || defined(OPENSSL_OPENBSD)
+
+// We assume that a method in this function is sufficient to detect fork events.
+// Returns 1 if a method sufficient for fork detection successfully
+// initializes. Otherwise returns 0.
+static int init_fork_detect_methods_sufficient(void *addr, long page_size) {
+
+  if (ignore_wipeonfork != 1 &&
+      init_fork_detect_wipeonfork(addr, page_size) == 1) {
+    return 1;
+  }
+
+  if (ignore_inheritzero != 1 &&
+      init_fork_detect_inheritzero(addr, page_size) == 1) {
+    return 1;
+  }
+
+  return 0;
+}
+
+// Best-effort attempt to initialize fork detection methods that provides
+// defense-in-depth. These methods should not be relied on for fork detection.
+// If initialization fails for one of these methods, just ignore it.
+static void init_fork_detect_methods_best_effort(void *addr, long page_size) {
+  // No methods yet. pthread_at_fork() would be a "best-effort" choice. It can
+  // detect fork events through e.g. fork(). But miss fork events through
+  // clone().
+}
+
 static void init_fork_detect(void) {
 
-  void *addr = NULL;
+  void *addr = MAP_FAILED;
+  long page_size = 0;
+
+  if (ignore_all_fork_detection() == 1) {
+    return;
+  }
 
-  // Check whether we are completely ignoring fork detection. This is only done
-  // during testing.
-  if (ignore_madv_wipeonfork == 1) {
+  page_size = sysconf(_SC_PAGESIZE);
+  if (page_size <= 0) {
     return;
   }
 
-  if (init_fork_detect_madv_wipeonfork(&addr) != 1) {
+  addr = mmap(NULL, (size_t)page_size, PROT_READ | PROT_WRITE,
+                    MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (addr == MAP_FAILED) {
```

Comment:
NP: You could check for `NULL` here, then not need the `GUARD_PTR` macro in the static functions above. (I don't think it's even possible for `mmap` to return NULL.)

Reply:
The guard is probably not necessary. But it was requested in a previous PR :)